packages/playwright-core/src/server/isomorphic/utilityScriptSerializers.ts

```diff
@@ -23,129 +23,142 @@ export type SerializedValue =
+  }
 
-export function parseEvaluationResultValue(value: SerializedValue, handles: any[] = []): any {
-  if (Object.is(value, undefined))
-    return undefined;
-  if (typeof value === 'object' && value) {
-    if ('v' in value) {
-      if (value.v === 'undefined')
-        return undefined;
-      if (value.v === 'null')
-        return null;
-      if (value.v === 'NaN')
-        return NaN;
-      if (value.v === 'Infinity')
-        return Infinity;
-      if (value.v === '-Infinity')
-        return -Infinity;
-      if (value.v === '-0')
-        return -0;
+  function isDate(obj: any): obj is Date {
+    return obj instanceof Date || Object.prototype.toString.call(obj) === '[object Date]';
+  }
+
+  function isError(obj: any): obj is Error {
+    return obj instanceof Error || (obj && obj.__proto__ && obj.__proto__.name === 'Error');
+  }
+
+  function parseEvaluationResultValue(value: SerializedValue, handles: any[] = []): any {
+    if (Object.is(value, undefined))
       return undefined;
+    if (typeof value === 'object' && value) {
+      if ('v' in value) {
+        if (value.v === 'undefined')
+          return undefined;
+        if (value.v === 'null')
+          return null;
+        if (value.v === 'NaN')
+          return NaN;
+        if (value.v === 'Infinity')
+          return Infinity;
+        if (value.v === '-Infinity')
+          return -Infinity;
+        if (value.v === '-0')
+          return -0;
+        return undefined;
+      }
+      if ('d' in value)
+        return new Date(value.d);
+      if ('r' in value)
+        return new RegExp(value.r.p, value.r.f);
+      if ('a' in value)
+        return value.a.map((a: any) => parseEvaluationResultValue(a, handles));
+      if ('o' in value) {
+        const result: any = {};
+        for (const { k, v } of value.o)
+          result[k] = parseEvaluationResultValue(v, handles);
+        return result;
+      }
+      if ('h' in value)
+        return handles[value.h];
     }
-    if ('d' in value)
-      return new Date(value.d);
-    if ('r' in value)
-      return new RegExp(value.r.p, value.r.f);
-    if ('a' in value)
-      return value.a.map((a: any) => parseEvaluationResultValue(a, handles));
-    if ('o' in value) {
-      const result: any = {};
-      for (const { k, v } of value.o)
-        result[k] = parseEvaluationResultValue(v, handles);
+    return value;
+  }
+
+  function serializeAsCallArgument(value: any, handleSerializer: (value: any) => HandleOrValue): SerializedValue {
+    return serialize(value, handleSerializer, new Set());
+  }
+
+  function serialize(value: any, handleSerializer: (value: any) => HandleOrValue, visited: Set<any>): SerializedValue {
+    const result = handleSerializer(value);
+    if ('fallThrough' in result)
+      value = result.fallThrough;
+    else
       return result;
+
+    if (visited.has(value)) {
+      if (ignoreCycles)
+        return '[Circular Ref]';
```

Comment:
- I am somewhat worried about performance, now that you can pass whatever to the binding.
- On the other hand, we can go further and restore cycles like we do with handles.

Reply:
Aliasing Window / Document / Node